fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,180 @@
+
+        int code = 0;
+        if (state.isCompressed()) {
+            code = code | ENCODING_COMPRESSED;
+        }
+        // Encryption will be supported in future
+        if (state.isEncrypted()) {
+            code = code | ENCODING_ENCRYPTED;
+        }
+        encoded[0] = (byte) code;
+        LOGGER.trace("Encoded data for compressionSupposed={}, encryptionSupposed={}, compressedEventually={}, encryptedEventually={}, originalDataSize={}, encodedDataSize={}",
+                compress, encrypt, state.isCompressed(), state.isEncrypted(), data.length, encoded.length);
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
+        final boolean encrypted = (encoding & ENCODING_ENCRYPTED) == ENCODING_ENCRYPTED;
+        final boolean compressed = (encoding & ENCODING_COMPRESSED) == ENCODING_COMPRESSED;
+
+        byte[] decoded = decompressIfNeeded(compressed, data, 1);
+        LOGGER.trace("Decoded data for compressed={}, encrypted={}, encodedDataSize={}, decodedDataSize={}",
+                compressed, encrypted, data.length, decoded.length);
+        return decoded;
+    }
+
+    @Nonnull
+    private static byte[] compressIfNeeded(boolean compressionNeeded, @Nonnull byte[] data, @Nonnull ByteBuffersDataOutput encodedDataOutput,
+                                           @Nonnull EncodingState state, int offset) {
+        if (!compressionNeeded) {
+            return fallBackToUncompressed(encodedDataOutput.toArrayCopy(), data, state, offset);
+        }
+
+        try (Compressor compressor = CompressionMode.HIGH_COMPRESSION.newCompressor()) {
+            encodedDataOutput.writeByte(COMPRESSION_VERSION_FOR_HIGH_COMPRESSION);
+            encodedDataOutput.writeVInt(data.length);
+            compressor.compress(data, 0, data.length, encodedDataOutput);
+            final byte[] compressedData = encodedDataOutput.toArrayCopy();
+
+            // Compress only if it helps to shorten the bytes
+            if (compressedData.length < data.length) {
+                state.setCompressed(true);
+                return compressedData;
+            } else {
+                return fallBackToUncompressed(compressedData, data, state, offset);
+            }
+        } catch (Exception e) {
+            throw new RecordCoreException("Lucene data compression failure", e);
+        }
+    }
+
+    private static byte[] fallBackToUncompressed(@Nonnull byte[] compressedData, @Nonnull byte[] originalData,
+                                                 @Nonnull EncodingState state, int offset) {
+        state.setCompressed(false);
+        final byte[] encoded = new byte[originalData.length + 1];
```

Comment:
Does this imply that `offset` is always 1 in this case? Otherwise, it would seem that one might try to copy `originalData` off the end of `encoded`.

It also seems like this is trying to add a byte with value 0 to the beginning of the new byte array, which I think is more easily done without the first arraycopy.

Reply:
Or maybe this is fixed by just making `originalData.length + 1` into `originalData.length + offset`

Reply:
Good catch! Yes this should be ```+ offset``` instead of ```+ 1```. Fixed it.